</div>


<a href="https://www.torogoz.club"><img src = "logo.png" width = 300, align = "center"></a>

<h1 align=center><font size = 5>Análisis de Datos con Python</font></h1>


<h1>Introducción</h1>
<h3>¡Bienvenido!</h3>

<p>
En esta sección aprenderá como acercarse a la adquisición de datos de varias formas y obtener información necesaria de un conjunto de datos. Al terminar este laboratorio podra cargar con éxito los datos al cuaderno de Jupyter y obtener aportaciones fundamentales mediante la libreria Pandas.
   </p>


<h1 id="data_acquisition">Obtención de Datos</h1>
<p>
Existen varios formatos para un conjunto de datos, .csv, .json, .xlsx, etc. Los datos pueden ser almacenados en distintos lugares, ya sea localmente o en línea.<b>
En estas sección aprenderá a cargar un conjunto de datos en su cuaderno de Jupyter.</b>
En nuestro caso el conjunto de datos Automobile es de una fuente en línea en formato CSV (valores separados por coma). Usemos este conjunto como ejemplo para practicar la lectura de datos.

<ul>
    <li>fuente de datos: <a href="https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ESedX20601300-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data</a></li>
    <li>tipo de datos: csv</li>
</ul>
La librería Pandas es una herramienta util que nos permite la lectura de varios conjuntos de datos en un dataframe; nuestra plataforma de cuaderno de Jupyter posee esta librería por defecto así que lo único que debemos hacer es importar Pandas sin necesidad de instalarlo.
</p>


In [8]:
# importar librería pandas 
import pandas as pd

<h2>Lectura de Datos</h2>
<p>
Utilizamos la función <code>pandas.read_csv()</code> para leer el archivo csv. Entre los corchetes pondremos la ruta del archivo sin comillas, de esta forma Pandas leerá el archivo en el dataframe desde esa dirección. La ruta del archivo puede ser tanto una URL como una dirección local.<br>
Debido a que los datos no tienen cabeceras podemos añadir un argumento <code>headers = None</code> dentro del método <code>read_csv()</code> para que Pandas no establezca de forma automática la primera fila como cabecera.<br>
Puede además asignar el conjunto de datos a cualquier variable que haya creado.

</p>


Este conjunto de datos fue almacenado como un objeto en IBM Cloud, haz clic <a href="https://cloud.ibm.com/registration/trial?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ESedX20601300-2021-01-01">AQUI</a> para almacenar gratuitamente.


In [9]:
# Importar la librería Pandas 
import pandas as pd

# Leer el archivo en línea desde la URL de arriba y asignarla a la variable "df"
other_path = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DA0101EN/auto.csv"
df = pd.read_csv(other_path, header=None)

Después de leer el conjunto de datos podemos utilizar el método <code>dataframe.head(n)</code> para revisar las primeras n filas del dataframe; donde n es u entero. Al contrario de <code>dataframe.head(n)</code>, <code>dataframe.tail(n)</code> mostrará las n filas del final del dataframe.


In [ ]:
# muestra las primeras 5 filas mediante el método dataframe.head()
print("The first 5 rows of the dataframe") 
df.head(5)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pregunta #1: </h1>
<b>Revisar las ultimas 10 filas del dataframe "df"</b>    
</div>


In [ ]:

print("10 ultimas") 
df.tail(10)

<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pregunta #1 Respuesta: </h1>
<b>Ejecute el código de abajo para ver la solución</b>    
</div>


Haz doble clic <b>aquí</b> para ver la solución.

<!-- The answer is below:

print("The last 10 rows of the dataframe\n")
df.tail(10)

-->


<h3>Añadir Cabeceras</h3>
<p>
Observe el conjunto de datos; Pandas automaticamente establece la cabecera en un entero a partir de 0.  
</p>
<p>
Para describir mejor nuestros datos podemos agregarle una cabecera, esta información esta disponible en:    
<a href="https://archive.ics.uci.edu/ml/datasets/Automobile?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDA0101ESedX20601300-2021-01-01" target="_blank">https://archive.ics.uci.edu/ml/datasets/Automobile</a>
</p>
<p>
De este modo debemos agregar las cabeceras manualmente.    
</p>
<p>
Primero creamos una lista "headers" que incluya todos los nombres de columna en orden.
Despues usamos <code>dataframe.columns = headers</code> para reemplazar las cabeceras por la lista que hemos creado.
</p>


In [ ]:
# crear la lista headers 
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
print("headers\n", headers)

Remplazamos las cabeceras y volvemos a revisar nuestro dataframe


In [17]:
headers = ["symboling","normalized-losses","make","fuel-type","aspiration", "num-of-doors","body-style",
         "drive-wheels","engine-location","wheel-base", "length","width","height","curb-weight","engine-type",
         "num-of-cylinders", "engine-size","fuel-system","bore","stroke","compression-ratio","horsepower",
         "peak-rpm","city-mpg","highway-mpg","price"]
df.columns = headers
df.head(10)

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   
5          2                 ?         audi       gas        std          two   
6          1               158         audi       gas        std         four   
7          1                 ?         audi       gas        std         four   
8          1               158         audi       gas      turbo         four   
9          0                 ?         audi       gas      turbo          two   

    body-style drive-wheels engine-location  wheel-base  ...  engine-size  \
0  convertible          rwd           front        88.6  ...          130   
1  convertible          rwd           front        88.6  ...          130   
2    hatchback          rwd           front        94.5  ...          152   
3        sedan          fwd           front        99.8  ...          109   
4        sedan          4wd           front        99.4  ...          136   
5        sedan          fwd           front        99.8  ...          136   
6        sedan          fwd           front       105.8  ...          136   
7        wagon          fwd           front       105.8  ...          136   
8        sedan          fwd           front       105.8  ...          131   
9    hatchback          4wd           front        99.5  ...          131   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19    3.40              10.0        102      5500       24   
4         mpfi  3.19    3.40               8.0        115      5500       18   
5         mpfi  3.19    3.40               8.5        110      5500       19   
6         mpfi  3.19    3.40               8.5        110      5500       19   
7         mpfi  3.19    3.40               8.5        110      5500       19   
8         mpfi  3.13    3.40               8.3        140      5500       17   
9         mpfi  3.13    3.40               7.0        160      5500       16   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  
5          25  15250  
6          25  17710  
7          25  18920  
8          20  23875  
9          22      ?  

[10 rows x 26 columns]

Podemos eliminar valores faltantes en la columna "price" de la siguiente manera


In [18]:
df.dropna(subset=["price"], axis=0)

symboling normalized-losses         make fuel-type aspiration  \
0            3                 ?  alfa-romero       gas        std   
1            3                 ?  alfa-romero       gas        std   
2            1                 ?  alfa-romero       gas        std   
3            2               164         audi       gas        std   
4            2               164         audi       gas        std   
..         ...               ...          ...       ...        ...   
200         -1                95        volvo       gas        std   
201         -1                95        volvo       gas      turbo   
202         -1                95        volvo       gas        std   
203         -1                95        volvo    diesel      turbo   
204         -1                95        volvo       gas      turbo   

    num-of-doors   body-style drive-wheels engine-location  wheel-base  ...  \
0            two  convertible          rwd           front        88.6  ...   
1            two  convertible          rwd           front        88.6  ...   
2            two    hatchback          rwd           front        94.5  ...   
3           four        sedan          fwd           front        99.8  ...   
4           four        sedan          4wd           front        99.4  ...   
..           ...          ...          ...             ...         ...  ...   
200         four        sedan          rwd           front       109.1  ...   
201         four        sedan          rwd           front       109.1  ...   
202         four        sedan          rwd           front       109.1  ...   
203         four        sedan          rwd           front       109.1  ...   
204         four        sedan          rwd           front       109.1  ...   

     engine-size  fuel-system  bore  stroke compression-ratio horsepower  \
0            130         mpfi  3.47    2.68               9.0        111   
1            130         mpfi  3.47    2.68               9.0        111   
2            152         mpfi  2.68    3.47               9.0        154   
3            109         mpfi  3.19    3.40              10.0        102   
4            136         mpfi  3.19    3.40               8.0        115   
..           ...          ...   ...     ...               ...        ...   
200          141         mpfi  3.78    3.15               9.5        114   
201          141         mpfi  3.78    3.15               8.7        160   
202          173         mpfi  3.58    2.87               8.8        134   
203          145          idi  3.01    3.40              23.0        106   
204          141         mpfi  3.78    3.15               9.5        114   

     peak-rpm city-mpg highway-mpg  price  
0        5000       21          27  13495  
1        5000       21          27  16500  
2        5000       19          26  16500  
3        5500       24          30  13950  
4        5500       18          22  17450  
..        ...      ...         ...    ...  
200      5400       23          28  16845  
201      5300       19          25  19045  
202      5500       18          23  21485  
203      4800       26          27  22470  
204      5400       19          25  22625  

[205 rows x 26 columns]

Ahora hemos leido con éxito el conjunto de datos y añadido las cabeceras correctas en el dataframe.


 <div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pregunta #2: </h1>
<b>Encuentre el nombre de las columnas del dataframe</b>
</div>


In [22]:
# Escribe tu código aquí y presiona Shift+Enter para ejecutar
print(df.columns)

Index(['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration',
       'num-of-doors', 'body-style', 'drive-wheels', 'engine-location',
       'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type',
       'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke',
       'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg',
       'highway-mpg', 'price'],
      dtype='object')


Haz doble clic <b>aquí</b> para ver la solución.

<!-- The answer is below:

print(df.columns)

-->


<h2>Guardar el Conjunto de Datos</h2>
<p>
De la misma forma, Pandas nos permite guardar el conjunto en formato CSV con el método <code>dataframe.to_csv()</code>, puede añadir la ruta al archivo y el nombre con comillas dentro de los corchetes.
</p>
<p>
    Por ejemplo, si guarda el dataframe <b>df</b> como <b>automobile.csv</b> en su equipo local, podría usar la sintaxis siguiente:
    </p>


Podemos leer y guardar con otros formatos y usar funciones similares a **`pd.read_csv()`** y **`df.to_csv()`** para otros formatos de datos, las funciones se muestran en la siguiente tabla:


<h2>Leer/Guardar en otros formatos</h2>

| Data Formate |        Read       |            Save |
| ------------ | :---------------: | --------------: |
| csv          |  `pd.read_csv()`  |   `df.to_csv()` |
| json         |  `pd.read_json()` |  `df.to_json()` |
| excel        | `pd.read_excel()` | `df.to_excel()` |
| hdf          |  `pd.read_hdf()`  |   `df.to_hdf()` |
| sql          |  `pd.read_sql()`  |   `df.to_sql()` |
| ...          |        ...        |             ... |


<h1 id="basic_insight">Características Básicas del Conjunto de Datos</h1>
<p>
Después de leer los datos en el dataframe Pandas es tiempo de explorar el conjunto de datos.<br>
Hay varias formas de obtener información esencial de los datos para entender mejor la totalidad de nuestro conjunto.</p>


<h2>Tipo de Datos</h2>
<p>
Los datos se encuentran en una variedad de tipos.<br>
Los tipos principales almacenados en dataframes de Pandas son <b>object</b>, <b>float</b>, <b>int</b>, <b>bool</b> y <b>datetime64</b>. Para aprender mejor acerca de cada atributo es mejor para nosotros saber el tipo de dato de cada columna. En Pandas:
</p>


In [ ]:
df.dtypes


Regresa una serie con el tipo de dato de cada columna.


In [ ]:
# revisar el tipo de dato del dataframe "df" con .dtypes
print(df.dtypes)

<p>
Como se muestra, se observa claramente que el tipo de dato de "symboling" y "curb-weight" es <code>int64</code>, "normalized-losses" es <code>object</code>, y "wheel-base" es <code>float64</code>, etc.
</p>
<p>
Estos tipos de datos pueden modificarse; aprenderemos a hacerlo en el siguiente módulo.    

</p>


<h2>Describe (método)</h2>
Si queremos obtener un resumen estadístico de cada columna, como el conteo, el valor medio de la columna, desviación estandar, etc. Utilizaremos el método describe:


Este método ofrece varios resumenes estadistícos, excluyendo valores  <code>NaN</code> (Not a Number).


In [ ]:
df.describe()

<p>
Este muestre el resumén estadístico de todas las columnas de tipo numerico (int, float).<br>
Por ejemplo, el atributo "symboling" tiene 205 recuentos, el valor medio de esta columna es 0.83, la desviación estandar es 1.25, el valor mínimo es -2, 25 percentil igual a 0, 50 percentil es 1, 75 percentil es 2 y el valor máximo es 3.
<br>
Sin embargo, si quisieramos además revisar todas las columnas incluyendo aquellas que son de tipo objeto, ¿Que tendríamos que hacer?  
<br><br>

Puede añadir un argumento <code>include = "all"</code> dentro del corchete. Hagámoslo de nuevo.

</p>


In [ ]:
# describe todas las columnas de "df" 
df.describe(include = "all")

<p>
Ahora se da el resumen estadistíco de todas las columnas, incluyendo los atributos de tipo objeto.<br>
Podemos observar ahora cuantos valores únicos, el cual es el valor mas alto y la frecuencia del valor alto en las columnas de tipo objeto.<br>
Algunos valores en la tabla de arriba se muestran como "NaN", esto se debe a que aquellos números no estan disponibles a pesar del tipo particular de columna.<br>    
</p>


<div class="alert alert-danger alertdanger" style="margin-top: 20px">
<h1> Pregunta #3: </h1>

<p>
Puede seleccionar las columna de un dataframe indicando el nombre de cada una, por ejemplo, puede seleccionar tres columnas de la siguiente manera:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3']]</code>
</p>
<p>
Donde "column" es el nombre de la columna se puede aplicar el método ".describe()" para obtener las estadísticas de aquellas columnas de la siguiente manera:
</p>
<p>
    <code>dataframe[[' column 1 ',column 2', 'column 3'] ].describe()</code>
</p>
Aplicar el método ".describe()" a las columnas 'length' y 'compression-ratio'. 
</div>


In [23]:
# Escribe tu código aquí y presiona Shift+Enter para ejecutar
df[['length', 'compression-ratio']].describe()

length  compression-ratio
count  205.000000         205.000000
mean   174.049268          10.142537
std     12.337289           3.972040
min    141.100000           7.000000
25%    166.300000           8.600000
50%    173.200000           9.000000
75%    183.100000           9.400000
max    208.100000          23.000000

Haz doble clic <b>aquí</b> para ver la solución.

<!-- The answer is below:

df[['length', 'compression-ratio']].describe()

-->


<h2>Info (método)</h2>
Otro método que puede usar para revisar el dataframe es:


Ofrece un resumen conciso del dataframe.


In [ ]:
# mirar la información de "df"
df.info

<p>
Podemos observar la información del dataframe con las primeras 30 filas y las últimas 30.
<br><br>
Además, tambien muestra que la totalidad del dataframe tiene 205 filas y 26 columnas en total.
</p>


<h1>¡Excelente! ¡Ha completado el Cuaderno de Introducción!</h1>
